src/core/shape_inference/include/broadcast_shape_inference.hpp

```diff
@@ -0,0 +1,301 @@
+                              " cannot be broadcasted (numpy mode) to ",
+                              target_shape[i],
+                              ". Allowed input dimension value would be 1",
+                              target_shape[i] != 1 ? " or " : "",
+                              target_shape[i] != 1 ? std::to_string(target_shape[i].get_length()) : "");
+    }
+}
+
+template <typename T>
+void set_result_shape_pdpd(const ov::Node* op,
+                           const T& arg0_shape,
+                           const T& target_shape,
+                           T& result_shape,
+                           const ov::op::BroadcastModeSpec& broadcast_spec) {
+    using DimType = typename std::iterator_traits<typename T::iterator>::value_type;
+    if (arg0_shape.rank().is_dynamic() || target_shape.rank().is_dynamic()) {
+        result_shape = PartialShape::dynamic(target_shape.rank());
+        return;
+    }
+    result_shape = target_shape;
+    auto& start_axis = broadcast_spec.m_axis;
+
+    NODE_VALIDATION_CHECK(op, start_axis >= 0, "Broadcast start_axis must be greater than 0");
+
+    for (size_t i = start_axis; i < target_shape.size(); i++) {
+        const auto& arg_dim = arg0_shape[i - start_axis];
+        if (arg_dim == 1) {
+            result_shape[i] = target_shape[i];
+        } else if (target_shape[i] == 1) {
+            result_shape[i] = arg_dim;
+        } else {
+            NODE_VALIDATION_CHECK(op,
+                                  DimType::merge(result_shape[i], arg_dim, target_shape[i]),
+                                  "Broadcast incorrect target shape. Expecting either 1 or ",
+                                  arg_dim,
+                                  " . Got ",
+                                  target_shape[i]);
+        }
+    }
+}
+
+template <typename T>
+void set_result_shape_bidirectional(const ov::Node* op, const T& arg_shape, T& target_shape, T& result_shape) {
+    if (arg_shape.rank().is_dynamic() || target_shape.rank().is_dynamic()) {
+        result_shape = PartialShape::dynamic();
+        return;
+    }
+    auto arg_shape_vec = arg_shape;
+
+    // Add left padding to shorter target or argument shape
+    const auto target_padded_rank = std::max(arg_shape_vec.size(), target_shape.size());
+    while (arg_shape_vec.size() < target_padded_rank) {
+        arg_shape_vec.insert(arg_shape_vec.begin(), 1);
+    }
+    while (target_shape.size() < target_padded_rank) {
+        target_shape.insert(target_shape.begin(), 1);
+    }
+
+    result_shape.resize(target_padded_rank);
+    for (size_t i = 0; i < target_shape.size(); ++i) {
+        if (arg_shape_vec[i] == 1) {
+            result_shape[i] = target_shape[i];
+        } else if (target_shape[i] == 1) {
+            result_shape[i] = arg_shape_vec[i];
+        } else {
+            NODE_VALIDATION_CHECK(op,
+                                  arg_shape_vec[i].compatible(target_shape[i]),
```

Comment:
how compatible is meaningful here? Example
```
arg_shape_vec[i] = 1
target_shape[i] = 2..5
```

Reply:
if `arg_shape_vec[i] = 1` code will run [this branch](https://github.com/openvinotoolkit/openvino/blob/8b4728383a3f34ae4f95af137fbd81b2ad1b22f2/src/core/shape_inference/include/broadcast_shape_inference.hpp#L148-L149) so result_shape[i] = 2..5
if both `arg_shape_vec[i] != 1` and `target_shape[i] != 1`  we will check whether they are compatible. Example
Compatible:
`arg_shape_vec[i] = 2..5
target_shape[i] = 3`
Not Compatible:
`arg_shape_vec[i] = 2..5
target_shape[i] = 6`

Reply:
I see that the code in broadcast changed a bit. Could you merge introduced logic here?

Reply:
Thank you Jane! I checked master branch, the only change is that bidirectional broadcast can propagate more accurate dynamic values when target_shape is dynamic. It's already supported in my PR, but the logic is a little different because I need to bypass MyriadX test case failure. But I see the test case is already remove in master. So I changed the code here to align with master. Please help review again. Thanks!